project_setup/docker/README.md

```diff
@@ -0,0 +1,264 @@
+# Docker for development
+
+Below we lay out recommendations for using Docker to wrap development
+dependencies. Unfortunately, this is a relatively new space and we don't have
+a consistent, end-to-end setup at the moment. Instead, we'll provide pointers,
+partial solutions and examples. We intend to flesh out complete solutions as
+they become apparent in addition to a comprehensive introduction.
+
+## Why?
+
+Consider your app's production environment. You're likely running on cloud.gov
+(which runs Ubuntu Linux) with specific versions of databases, app language
+(e.g. Go/Node/Python/Ruby), and required libraries (you are [pinning your
+dependencies](https://pages.18f.gov/before-you-ship/infrastructure/pinning-dependencies/),
+right?). Now consider your local environment; if you're an 18F engineer, you
+likely have OS X, language version management tools, and an assortment of
+strategies for running databases, worker queues, etc. If you're a designer,
+open source contributor, or working at a partner agency, we should assume even
+less. Not only do we have significant mismatches between these environments,
+we also require significant setup for new and infrequent contributors.
+
+Docker promises a partial solution to this problem by wrapping dependencies
+into a consistent, reproducible environment. While we don't generally support
+Docker in production, we can create a setup that matches cloud.gov relatively
+closely and which makes running our app painless.
+
+## Recommendations
+
+We hope to have an end-to-end recommendation in the future, but for now we
+provide these scoped recommendations.
+
+### Use Docker-Compose
+
+Docker introductions tend to highlight the
+[`Dockerfile`](https://docs.docker.com/engine/reference/builder/), a
+procedural description of how to build a Docker image, but that's generally
+not encompassing enough for development needs. We will likely want to run a
+database (or other service) separate from our application, and provide
+different environments (e.g. one for building, one for development, one for
+production-like); while it's possible to include all of that in a single
+image, it's generally better to use separate, focused images. Moreover, while
+Dockerfiles provide _hints_ around which ports should be exposed, which files
+should be shared with the host, etc. While we're running the app, we want to
+define _specific_ values.
+
+[Docker-compose files](https://docs.docker.com/compose/compose-file/) solve
+both of these problems by allowing us to specify different services,
+dependencies between those services, and specific values for exposed ports,
+mount points, etc. Though it feels more complex at first, we recommend using
+docker-compose even when a single Dockerfile would suffice so that we have a
+consistent tool across 18F engineering.
```

Comment:
One of the big "a ha" moments for me in my use of docker was transitioning from thinking of docker as a VM (and consequently to `Dockerfile` as a build script for a VM) to docker as an orchestration of containers, each of which performs one service, the latter of which is made practically effortless by docker-compose. dunno if that's worth highlighting here, but figured i'd mention it.